org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/IfStatement.java

```diff
@@ -300,37 +300,30 @@ public StringBuilder printStatement(int indent, StringBuilder output) {
 	}
 	return output;
 }
-private void resolveIfStatement(BlockScope scope) {
-	TypeBinding type = this.condition.resolveTypeExpecting(scope, TypeBinding.BOOLEAN);
-	this.condition.computeConversion(scope, type, type);
-	if (this.thenStatement != null)
-		this.thenStatement.resolve(scope);
-	if (this.elseStatement != null)
-		this.elseStatement.resolve(scope);
+
+@Override
+public LocalVariableBinding[] getPatternVariablesLiveUponCompletion() {
+	if (!this.condition.containsPatternVariable() || doesNotCompleteNormally())
+		return NO_VARIABLES;
+	if (this.thenStatement != null && this.thenStatement.doesNotCompleteNormally())
+		return this.condition.getPatternVariablesWhenFalse();
+	if (this.elseStatement != null && this.elseStatement.doesNotCompleteNormally())
+		return this.condition.getPatternVariablesWhenTrue();
+	return NO_VARIABLES;
 }
 @Override
 public void resolve(BlockScope scope) {
-	if (containsPatternVariable()) {
-		this.condition.collectPatternVariablesToScope(null, scope);
-		LocalVariableBinding[] patternVariablesInTrueScope = this.condition.getPatternVariablesWhenTrue();
-		LocalVariableBinding[] patternVariablesInFalseScope = this.condition.getPatternVariablesWhenFalse();
-		TypeBinding type = this.condition.resolveTypeExpecting(scope, TypeBinding.BOOLEAN);
-		this.condition.computeConversion(scope, type, type);
+	this.condition.collectPatternVariablesToScope(null, scope);
+	LocalVariableBinding[] patternVariablesInTrueScope = this.condition.getPatternVariablesWhenTrue();
+	LocalVariableBinding[] patternVariablesInFalseScope = this.condition.getPatternVariablesWhenFalse();
+	TypeBinding type = this.condition.resolveTypeExpecting(scope, TypeBinding.BOOLEAN);
+	this.condition.computeConversion(scope, type, type);
 
-		if (this.thenStatement != null) {
-			this.thenStatement.resolveWithPatternVariablesInScope(patternVariablesInTrueScope, scope);
-		}
-		if (this.elseStatement != null) {
-			this.elseStatement.resolveWithPatternVariablesInScope(patternVariablesInFalseScope, scope);
-		}
-		if (this.thenStatement != null)
-			this.thenStatement.promotePatternVariablesIfApplicable(patternVariablesInFalseScope,
-				this.thenStatement::doesNotCompleteNormally);
-		if (this.elseStatement != null)
-			this.elseStatement.promotePatternVariablesIfApplicable(patternVariablesInTrueScope,
```

Comment:
@jarthana - This can be the starting point for the review. An `If_Statement` has only too localized context to decide if some variables should be "promoted" into the scope. 

For example the if statement in **bold** below **should not promote** its binding variable:


if (obj instanceof Integer r) {
    System.out.println();
} else **if (obj instanceof Double c) {
    System.out.println();
} else {
    throw new IllegalArgumentException("invalid type"); // works OK without this line
}**

// Control can reach here without 'c' being definitely assigned! */

The if statement in bold does not know if it is a part of a chain of if-else-if statements. So we move the decision to a different place - the newly invented method `ASTNode.resolveStatements`